packages/use-dataloader/src/useDataLoader.ts

```diff
@@ -1,171 +1,126 @@
-    [request.status],
-  )
-  const isError = useMemo(
-    () => request.status === StatusEnum.ERROR,
-    [request.status],
-  )
-  const isPolling = useMemo(
-    () => !!(enabled && pollingInterval && (isSuccess || isLoading)),
-    [isSuccess, isLoading, enabled, pollingInterval],
+      request.load(true).then(onSuccessRef.current).catch(onErrorRef.current),
+    [request],
   )
 
   useEffect(() => {
-    if (enabled) {
-      // launch should never throw
-      // eslint-disable-next-line @typescript-eslint/no-floating-promises
-      request.load()
-    }
-  }, [request, enabled, isIdle])
+    request.method = method
+  }, [method, request])
 
   useEffect(() => {
-    if (pollingInterval !== request.pollingInterval) {
-      request.setPollingInterval(pollingInterval)
-    }
-  }, [pollingInterval, request])
+    onSuccessRef.current = onSuccess
+  }, [onSuccess])
 
   useEffect(() => {
-    if (needPolling !== request.needPolling) {
-      request.setNeedPolling(needPolling ?? true)
-    }
-  }, [needPolling, request])
+    onErrorRef.current = onError ?? onGlobalError
+  }, [onError, onGlobalError])
 
   useEffect(() => {
-    isFetchingRef.current = isLoading || isPolling
-  }, [isLoading, isPolling])
-
-  useLayoutEffect(() => {
-    cancelMethodRef.current = request?.cancel
-  }, [request?.cancel])
+    if (enabled && request.loadCount === 0) {
+      if (keepPreviousData) {
+        previousDataRef.current = request.data
+      }
+      request.load().then(onSuccessRef.current).catch(onErrorRef.current)
+    }
+  }, [enabled, request, keepPreviousData])
 
   useEffect(() => {
-    isMountedRef.current = true
-
-    return () => {
-      isMountedRef.current = false
-      if (isFetchingRef.current && cancelMethodRef.current) {
-        // eslint-disable-next-line @typescript-eslint/no-floating-promises
-        cancelMethodRef.current()
+    let timeout: NodeJS.Timeout
+
+    if (
+      pollingInterval &&
+      needPolling &&
+      (request.status === StatusEnum.SUCCESS ||
+        request.status === StatusEnum.ERROR)
+    ) {
+      if (
+        (typeof needPolling === 'function' && needPolling(request.data)) ||
+        (typeof needPolling !== 'function' && needPolling)
+      ) {
+        timeout = setTimeout(() => {
+          request
+            .load(true)
+            .then(onSuccessRef.current)
+            .catch(onErrorRef.current)
+        }, pollingInterval)
       }
-      unsubscribeRequestRef.current?.()
     }
-  }, [])
 
-  useEffect(() => () => {
-    if (request.getData() !== previousDataRef.current && keepPreviousData) {
-      previousDataRef.current = request.getData()
+    return () => {
+      if (timeout) clearTimeout(timeout)
     }
-  })
+  }, [pollingInterval, needPolling, request, request.status, request.data])
 
   return {
-    data: request.getData() ?? (initialData as T),
-    error: request?.error,
+    data: !request.isFirstLoading ? request.data : initialData,
```

Comment:
I have a tiny improvement here, lets say your initialData is a complex object wouldn't it be nice to spread its default values into data ?

Something like this when request.data and initialData are both objects:
```ts
data: !request.isFirstLoading ? { ...initialData, ...request.data } : initialData,
```